h2/src/main/org/h2/value/TypeInfo.java

```diff
@@ -824,6 +824,87 @@ private static boolean areComparable(TypeInfo t1, TypeInfo t2) {
         return false;
     }
 
+    /**
+     * Determines whether two specified types have the same ordering rules.
+     *
+     * @param t1
+     *            first data type
+     * @param t2
+     *            second data type
+     * @return whether types are comparable
+     */
+    public static boolean haveSameOrdering(TypeInfo t1, TypeInfo t2) {
+        int vt1 = (t1 = t1.unwrapRow()).getValueType(), vt2 = (t2 = t2.unwrapRow()).getValueType();
+        if (vt1 > vt2) {
+            int vt = vt1;
+            vt1 = vt2;
+            vt2 = vt;
+            TypeInfo t = t1;
+            t1 = t2;
+            t2 = t;
+        }
+        if (vt1 <= Value.NULL) {
+            return true;
+        }
+        if (vt1 == vt2) {
+            switch (vt1) {
+            case Value.ARRAY:
+                return haveSameOrdering((TypeInfo) t1.getExtTypeInfo(), (TypeInfo) t2.getExtTypeInfo());
+            case Value.ROW: {
+                Set<Entry<String, TypeInfo>> f1 = ((ExtTypeInfoRow) t1.getExtTypeInfo()).getFields();
+                Set<Entry<String, TypeInfo>> f2 = ((ExtTypeInfoRow) t2.getExtTypeInfo()).getFields();
+                int degree = f1.size();
+                if (f2.size() != degree) {
+                    return false;
+                }
```

Comment:
What about if col has type
   ROW(VARCHAR, VARCHAR, INT)
and query has type
   ROW(VARCHAR, VARCHAR)
then the index would still be usable because the col ordering requirements are strictly a superset of the query ordering requirements
?

Reply:
Row values of different degree can't be compared with each other. This condition exists here only for safety, such comparisons are normally rejected earlier.

Reply:
Only arrays with different cardinality are comparable (if types of elements are comparable).